clients/Android/ultrablue/app/src/main/java/fr/gouv/ssi/ultrablue/fragments/ProtocolFragment.kt

```diff
@@ -1,43 +1,420 @@
+            about chunking them.
+         */
+        override fun onCharacteristicWrite(gatt: BluetoothGatt?, characteristic: BluetoothGattCharacteristic?, status: Int) {
+            super.onCharacteristicWrite(gatt, characteristic, status)
+            protocol?.let { proto ->
+                characteristic?.let {
+                    proto.onMessageWrite()
+                }
+            }
+        }
+    }
+
+    /*
+        Fragment lifecycle methods:
+     */
 
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
-        setHasOptionsMenu(true)
-        device = requireArguments().getSerializable("device") as Device?
-        return inflater.inflate(R.layout.fragment_device_list, container, false)
+        val menuHost = requireActivity()
+        menuHost.addMenuProvider(object: MenuProvider {
+            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
+                menuInflater.inflate(R.menu.action_bar, menu)
+            }
+            override fun onPrepareMenu(menu: Menu) {
+                super.onPrepareMenu(menu)
+                menu.findItem(R.id.action_edit).isVisible = false
+                menu.findItem(R.id.action_add).isVisible = false
+            }
+            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
+                return false
+            }
+        })
+        (activity as MainActivity).showUpButton()
+        return inflater.inflate(R.layout.fragment_protocol, container, false)
     }
 
+    @SuppressLint("MissingPermission")
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-        (requireActivity() as MainActivity).supportActionBar?.title = if (device != null) {
-            "Attestation in progress"
+		val device = requireArguments().getSerializable("device") as Device
+        if (device.name.isEmpty()) {
+            state = State.ENROLLMENT
+            activity?.title = "Enrollment in progress"
+        } else {
+            state = State.AUTHENTICATION
+            activity?.title = "Attestation in progress"
+        }
+        logger = Logger(activity as MainActivity?, view.findViewById(R.id.logger_text_view), onError = {
+            // TODO: Ask the user if he wants to inspect the logs or go back to the menu
+        })
+
+        if (!hasBluetoothPermission()) {
+            return
+        }
+
+        /*
+            As the following operations are not blocking, we let them run and give them
+            a callback to call when completed.
+            Briefly, those operations start Bluetooth, search for the attesting device,
+            connect to it, and then start the Ultrablue Protocol.
+
+            We deliberately break indentation to make this callback-based code read like
+            direct style.
+         */
+        askForLocationPermission(onSuccess = {
```

Comment:
Rename askForLocationPermission to askForPermission, taking a permission as a parameter (or take an array of permissions).

Ask for BLUETOOTH_CONNECT and BLUETOOTH_SCAN: https://developer.android.com/guide/topics/connectivity/bluetooth/permissions